### PR TITLE
Update React development stack

### PR DESCRIPTION
**Pull request by watsonbox**
The goal here is to update the project to a more modern React stack, brining several benefits:

- Using [`create-react-app`](https://create-react-app.dev/) is an established standard development stack which should make contributing easier
- Build step separates production and development environments (file structure, tests, performance)
- More robust development toolchain including live reload, auto test runs, linting etc
- Much improved testing framework (JEST)
- Up to date versions of React, Font Awesome, JSZip and FileSaver.js
- Sass for styling
- Clear path to full ES6 / TypeScript

### Out of Scope

I've deliberately omitted some things in this step, in order to keep the scope under control:

- Bootstrap upgrade
- Removal of jQuery
- Migration to TypeScript
- New features

### CODE REVIEW

File: src/components/PlaylistExporter.jsx [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+    }.bind(this))
+  },
+
+  csvData: function(access_token, playlist) {
+    var requests = [];
+    var limit = playlist.tracks.limit || 100;
+
+    for (var offset = 0; offset < playlist.tracks.total; offset = offset + limit) {
+      requests.push(
+        apiCall(playlist.tracks.href.split('?')[0] + '?offset=' + offset + '&limit=' + limit, access_token)
```
Comment (tesnouf-drivy) [inline src/components/PlaylistExporter.jsx:21]: you should use a literals template for this case:
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Template_literals

```ts
apiCall(`${playlist.tracks.href.split('?')[0]}?offset=${offset}&limit=${limit}`, access_token)
```

File: src/components/PlaylistExporter.jsx [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+        "Added By",
+        "Added At"
+      ]);
+
+      let csvContent = '';
+
+      tracks.forEach(function(row, index){
+        let dataString = row.map(function (cell) { return '"' + String(cell).replace(/"/g, '""') + '"'; }).join(",");
+        csvContent += dataString + "\n";
+      });
```
Comment (tesnouf-drivy) [inline src/components/PlaylistExporter.jsx:77]: When iterating on a loop to build something, you could use the `reduce` function

```ts
const csvContent = tracks.reduce( (acc, row) => {
return `${acc}${row.map(cell => '"' + String(cell).replace(/"/g, '""') + '"'; }).join(","))}\n`
      }, '');
```

File: src/components/PlaylistRow.jsx [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+  exportPlaylist() {
+    PlaylistExporter.export(this.props.access_token, this.props.playlist);
+  }
+
+  renderTickCross(condition) {
+    if (condition) {
+      return <FontAwesomeIcon icon={['far', 'check-circle']} size="sm" />
+    } else {
+      return <FontAwesomeIcon icon={['far', 'times-circle']} size="sm" style={{ color: '#ECEBE8' }} />
+    }
```
Comment (tesnouf-drivy) [inline src/components/PlaylistRow.jsx:16]: for simple if/else, the ternary condition is frequently used but it is a matter of taste

```ts
return condition ?
   <FontAwesomeIcon icon={['far', 'check-circle']} size="sm" /> :
   <FontAwesomeIcon icon={['far', 'times-circle']} size="sm" style={{ color: '#ECEBE8' }} />
```

File: src/components/PlaylistRow.jsx [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+  }
+
+  renderIcon(playlist) {
+    if (playlist.name === 'Liked') {
+      return <FontAwesomeIcon icon={['far', 'heart']} style={{ color: 'red' }} />;
+    } else {
+      return <FontAwesomeIcon icon={['fas', 'music']} />;
+    }
+  }
```
Comment (tesnouf-drivy) [inline src/components/PlaylistRow.jsx:25]: if you rely only on 1 property of your `playlist` object, maybe it should be your default parameter

```ts
renderIcon(playlistName) {
  return playlistName === 'Liked' ?
   <FontAwesomeIcon icon={['far', 'heart']} style={{ color: 'red' }} /> : 
   <FontAwesomeIcon icon={['fas', 'music']} />
```

File: src/components/PlaylistRow.jsx [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+
+    return (
+      <tr key={playlist.uri}>
+        <td>{this.renderIcon(playlist)}</td>
+        <td><a href={playlist.uri}>{playlist.name}</a></td>
+        <td><a href={playlist.owner.uri}>{playlist.owner.display_name}</a></td>
+        <td>{playlist.tracks.total}</td>
+        <td>{this.renderTickCross(playlist.public)}</td>
+        <td>{this.renderTickCross(playlist.collaborative)}</td>
+        <td className="text-right"><button className="btn btn-default btn-xs btn-success" type="submit" onClick={this.exportPlaylist.bind(this)}><span className="glyphicon glyphicon-save"></span> Export</button></td>
```
Comment (tesnouf-drivy) [inline src/components/PlaylistRow.jsx:49]: when using React class component, you're supposed to bind this into the constructor:
https://www.digitalocean.com/community/tutorials/react-constructors-with-react-components
`Bind events all in one place` section

File: src/components/Paginator.jsx [hunk 6]
```diff
@@ -0,0 +1,46 @@
+import React from "react"
+
+class Paginator extends React.Component {
```
Comment (tesnouf-drivy) [inline src/components/Paginator.jsx:3]: constructor is missing:
https://www.digitalocean.com/community/tutorials/react-constructors-with-react-components

File: src/components/PlaylistTable.jsx [hunk 7]
```diff
@@ -0,0 +1,139 @@
...
+
+class PlaylistTable extends React.Component {
+  state = {
+    playlists: [],
+    playlistCount: 0,
+    likedSongsLimit: 0,
+    likedSongsCount: 0,
+    nextURL: null,
+    prevURL: null
+  }
```
Comment (tesnouf-drivy) [inline src/components/PlaylistTable.jsx:18]: should be defined in your constructor:
https://reactjs.org/docs/react-component.html#constructor

File: src/helpers.jsx [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+  }).fail(function (jqXHR, textStatus) {
+    if (jqXHR.status === 401) {
+      // Return to home page after auth token expiry
+      window.location.href = window.location.href.split('#')[0]
+    } else if (jqXHR.status === 429) {
+      // API Rate-limiting encountered
+      window.location.href = window.location.href.split('#')[0] + '?rate_limit_message=true'
+    } else {
+      // Otherwise report the error so user can raise an issue
+      alert(jqXHR.responseText);
```
Comment (tesnouf-drivy) [inline src/helpers.jsx:41]: I guess it's a critical case...

File: src/components/Paginator.jsx [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+              { /* eslint-disable-next-line  */ }
+              <a href="#" aria-label="Next" onClick={this.nextClick.bind(this)}>
+                <span aria-hidden="true">&raquo;</span>
+              </a>
+            </li>
+          </ul>
+        </nav>
+      )
+    } else {
+      return <div>&nbsp;</div>
```
Comment (tesnouf-drivy) [inline src/components/Paginator.jsx:41]: it's a little bit odd to return an empty `div`
Could `return null` do the job ? or at least `return <div></div>` ?

File: src/components/PlaylistTable.jsx [hunk 10]
```diff
@@ -0,0 +1,139 @@
+import React from "react"
+import $ from "jquery" // TODO: Remove jQuery dependency
```
Comment (tesnouf-drivy) [inline src/components/PlaylistTable.jsx:2]: your comment say it all ;)

File: src/components/PlaylistTable.jsx [hunk 11]
```diff
@@ -0,0 +1,139 @@
...
+          <table className="table table-hover">
+            <thead>
+              <tr>
+                <th style={{width: "30px"}}></th>
+                <th>Name</th>
+                <th style={{width: "150px"}}>Owner</th>
+                <th style={{width: "100px"}}>Tracks</th>
+                <th style={{width: "120px"}}>Public?</th>
+                <th style={{width: "120px"}}>Collaborative?</th>
+                <th style={{width: "100px"}} className="text-right">
```
Comment (tesnouf-drivy) [inline src/components/PlaylistTable.jsx:117]: you should define your width using css classes.
Directly defining the style object is required when using variable / computing the value.
Aside of separating the visual from your logic, doing so won't break your Jest snapshots if you decide to change a column width

**Review by tesnouf-drivy — COMMENTED** [review]
Some minors comments/suggestions:

Some code could be more ES6+ friendly and React Components are a little bit legacy (Class Component instead of functional) and constructors are missing

I didn't focus on the logical side but this little project is neat !

File: src/components/Login.jsx [hunk 12]
```diff
@@ -0,0 +1,16 @@
+import React from "react"
+import { Button } from "react-bootstrap"
+import { FontAwesomeIcon } from "@fortawesome/react-fontawesome"
+import { authorize } from "helpers"
+
+class Login extends React.Component {
```
[the remark this answers is not in the document]
Comment (watsonbox) [reply src/components/Login.jsx:6]: Okay noted, I may simplify these then 👍

File: src/components/PlaylistExporter.jsx [hunk 13]
```diff
@@ -0,0 +1,88 @@
+import $ from "jquery" // TODO: Remove jQuery dependency
+import { saveAs } from "file-saver"
+
+import { apiCall } from "helpers"
+
+// Handles exporting a single playlist as a CSV file
+var PlaylistExporter = {
```
[the remark this answers is not in the document]
Comment (watsonbox) [reply src/components/PlaylistExporter.jsx:7]: Yep I think I'll clean up all the legacy code file by file once this is merged 👍

File: src/components/PlaylistExporter.jsx [hunk 14]
```diff
@@ -0,0 +1,88 @@
...
+  export: function(access_token, playlist) {
+    this.csvData(access_token, playlist).then(function(data) {
+      var blob = new Blob([ data ], { type: "text/csv;charset=utf-8" });
+      saveAs(blob, this.fileName(playlist), true);
+    }.bind(this))
+  },
```
[the remark this answers is not in the document]
Comment (watsonbox) [reply src/components/PlaylistExporter.jsx:13]: Interesting, I wondered which approach was more idiomatic these days 👌

File: src/components/PlaylistTable.jsx (hunk 7 above)
Comment (watsonbox) [reply src/components/PlaylistTable.jsx:18]: I read that if the constructor isn't required for anything else, this is cleaner e.g.:

https://dev.to/ascorbic/class-fields-are-coming-heres-what-that-means-for-react--3a87

File: src/components/Paginator.jsx (hunk 6 above)
Comment (watsonbox) [reply src/components/Paginator.jsx:3]: > The constructor comes in very handy, but sometimes it’s just too much. If you aren’t planning on maintaining state in your component and aren’t binding any event handlers, you don’t need to define a constructor at all.

So I agree, this one should have a constructor for binding the event handlers 👍

File: src/helpers.jsx (hunk 8 above)
Comment (watsonbox) [reply src/helpers.jsx:41]: Meh, again I think this should be improved at some point.

File: src/components/PlaylistTable.jsx (hunk 11 above)
Comment (watsonbox) [reply src/components/PlaylistTable.jsx:117]: Very good point about the Jest snapshots! Indeed, something else to add to the cleanup list 👍

**Comment by watsonbox** [comment]
Thanks a lot for the comments @tesnouf-drivy 🙏

**Comment by pavel-aicradle** [comment]
My PR gets rid of all jquery. Not sure how my React is compared to this PR; it is the first and only thing I've ever done in React. Honestly don't love React. Super verbose and annoying.

**Pull request merged**